elastic/elastic-agent
# Integration tests autodiscovery

## What does this PR do?
Implement integration test autodiscovery using a plain `go test` call like `go test -tags integration github.com/elastic/elastic-agent/testing/integration -v -args -integration.dry-run=true -integration.autodiscover`.

The YAML output is written in a file (by default `discovered_tests.yaml`) and is structured  like the sample below (some parts have been omitted to keep the sample short)
```yaml
- os_family: windows    # This indicates on which OS type we need to run the test
  arch: amd64                 # This indicates on which CPU architecture we need to run the test
  os: ""                              # This indicates on which OS (by name) we need to run the test ("" means any OS for the given type)
  version: ""                     # This indicates on which OS version we need to run the test ("" means any version for the given type)
  groups:                          # List of test groups to run on the runner
    - name: upgrade        # Test group name
      tests:                           # List of tests belonging to the group
        - name: TestStandaloneUpgradeRollback # Test name
          metadata:
            local: false            # True if the test can be run locally
            sudo: true             # True if the test requires sudo
        - name: TestStandaloneUpgradeRollbackOnRestarts ...

- os_family: linux
  arch: amd64
  os: ubuntu
  version: ""
  groups:
    - name: deb
      tests:
        - name: TestDebLogIngestFleetManaged
          metadata:
            local: false
            sudo: true
        - name: TestDebFleetUpgrade
          metadata:
            local: false
            sudo: true
- os_family: linux
  arch: amd64
  os: rhel
  version: ""
  groups:
    - name: rpm
      tests:
        - name: TestRpmLogIngestFleetManaged
          metadata:
            local: false
            sudo: true
        - name: TestRpmFleetUpgrade
          metadata:
            local: false
            sudo: true
```

In order to filter the output we can use a tool like yq

```shell
➜  elastic-agent git:(integration-tests-autodiscovery) ✗ yq '[.[].groups.[].name] | unique' testing/integration/discovered_tests.yaml
```

![image](https://github.com/user-attachments/assets/1caa205e-1882-465c-bde9-a8330725210e)



## Why is it important?

This is necessary to dynamically generate CI steps or provisioning for running the tests on the appropriate platforms



## Checklist




## Disruptive User Impact



## How to test this PR locally

In order to generate the YAML file we can run
```shell
go test -tags integration github.com/elastic/elastic-agent/testing/integration -v -args -integration.dry-run=true -integration.autodiscover
```
this will generate a file in `<repo root>/testing/integration/discovered_tests.yaml`

It can be combined with the existing group and sudo filters (see `-integration.groups` and `-integration.sudo`)



## Related issues


-

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
